In [6]:
pip install nbformat -q

Note: you may need to restart the kernel to use updated packages.


In [2]:
pip install gudhi -q

Note: you may need to restart the kernel to use updated packages.


In [1]:
from transformers import GPT2Tokenizer, GPT2Model

# Initialize the model and tokenizer
tokenizer = GPT2Tokenizer.from_pretrained('gpt2')
model = GPT2Model.from_pretrained('gpt2')

sentences = ["Quantum information theory is interesting.", 
             "Quantum information theory allows us to study attention using entanglement."]

# We need to find the context vectors for "Quantum" in each sentence
for sentence in sentences:
    # Tokenize the sentence
    inputs = tokenizer(sentence, return_tensors='pt')

    # Pass the tokens through the model to get hidden states
    outputs = model(**inputs)

    # outputs[0] contains the hidden states (context vectors) for each token
    hidden_states = outputs[0]

    # Find the index of "Quantum" in the tokenized sentence
    quantum_index = inputs['input_ids'][0].tolist().index(tokenizer.encode('Quantum')[0])

    # Extract the context vector for "Quantum"
    quantum_context_vector = hidden_states[0, quantum_index, :]

    print(f'Context vector for "Quantum" in sentence "{sentence}": {quantum_context_vector}')


Context vector for "Quantum" in sentence "Quantum information theory is interesting.": tensor([-2.0688e-01, -1.9257e-01, -3.0199e-01, -3.0512e-02,  1.4486e-02,
        -8.1787e-02, -1.2394e+00, -5.7916e-02, -3.3286e-01,  8.1929e-02,
         4.5471e-01,  1.1390e-02, -3.8550e-03,  8.7279e-02, -2.7556e-01,
         4.7814e-02, -3.0552e-01, -1.3261e-01,  8.0164e-02, -1.2619e-01,
         1.9125e-01, -1.8217e-01, -1.7503e-01,  1.8931e-01,  4.5310e-02,
         1.8951e-01, -1.7572e-01, -2.7025e-01, -1.1635e-01, -4.6141e-01,
         4.0340e-02,  6.6976e-02, -4.5858e-02, -2.2176e-01, -1.5579e-01,
         1.6255e-01,  1.9813e+01, -1.4622e-01, -4.1514e-02,  4.4666e-02,
        -1.6562e-02,  7.3794e-02, -2.5414e-02, -3.6816e-01, -1.5535e-01,
         3.1343e-02, -7.5143e-02,  1.5369e-02, -3.4010e-01, -1.7994e-01,
        -2.0722e-01,  1.5461e-01, -1.5441e-01, -6.9989e-02, -1.1230e-01,
        -2.1122e-01, -1.7243e-01, -3.1767e-01, -1.5880e-01, -6.3923e-02,
         1.8955e-01,  3.3378e-02, -1.

In [4]:
import torch
from transformers import GPT2Tokenizer, GPT2Model

# Initialize the model and tokenizer
tokenizer = GPT2Tokenizer.from_pretrained('gpt2')
model = GPT2Model.from_pretrained('gpt2')

sentences = ["Quantum information theory is interesting.", 
             "Quantum information theory allows us to study attention using entanglement."]

# Function to calculate Euclidean distance
def euclidean_distance(a, b):
    return torch.dist(a, b)

# Iterate over sentences
for sentence in sentences:
    # Tokenize the sentence
    inputs = tokenizer(sentence, return_tensors='pt')

    # Pass the tokens through the model to get hidden states
    outputs = model(**inputs)

    # outputs[0] contains the hidden states (context vectors) for each token
    hidden_states = outputs[0]

    # Tokenize the phrase separately to know how many tokens it contains
    phrase_tokens = tokenizer.encode("Quantum information theory")

    # Find the starting index of "Quantum information theory" in the tokenized sentence
    start_index = inputs['input_ids'][0].tolist().index(phrase_tokens[0])

    # Extract the context vectors for the phrase "Quantum information theory"
    phrase_context_vectors = hidden_states[0, start_index : start_index + len(phrase_tokens), :]

    # Compute pairwise Euclidean distances and store in a matrix
    distance_matrix = torch.zeros(len(phrase_tokens), len(phrase_tokens))
    for i in range(len(phrase_tokens)):
        for j in range(len(phrase_tokens)):
            distance_matrix[i, j] = euclidean_distance(phrase_context_vectors[i], phrase_context_vectors[j])

    print(f'Distance matrix for "Quantum information theory" in sentence "{sentence}":\n {distance_matrix}\n')


Distance matrix for "Quantum information theory" in sentence "Quantum information theory is interesting.":
 tensor([[  0.0000, 119.8416, 172.8208, 166.4057],
        [119.8416,   0.0000,  58.0346,  55.0810],
        [172.8208,  58.0346,   0.0000,  25.0608],
        [166.4057,  55.0810,  25.0608,   0.0000]], grad_fn=<CopySlices>)

Distance matrix for "Quantum information theory" in sentence "Quantum information theory allows us to study attention using entanglement.":
 tensor([[  0.0000, 119.8416, 172.8207, 166.4057],
        [119.8416,   0.0000,  58.0346,  55.0810],
        [172.8207,  58.0346,   0.0000,  25.0608],
        [166.4057,  55.0810,  25.0608,   0.0000]], grad_fn=<CopySlices>)



In [1]:
import torch
from transformers import GPT2Tokenizer, GPT2Model
from sklearn.decomposition import PCA
import plotly.graph_objects as go

# Initialize the model and tokenizer
tokenizer = GPT2Tokenizer.from_pretrained('gpt2')
model = GPT2Model.from_pretrained('gpt2')

sentences = ["Quantum information theory is interesting.", 
             "Quantum information theory allows us to study attention using entanglement."]

# Function to calculate Euclidean distance
def euclidean_distance(a, b):
    return torch.dist(a, b)

# Iterate over sentences
for sentence in sentences:
    # Tokenize the sentence
    inputs = tokenizer(sentence, return_tensors='pt')

    # Pass the tokens through the model to get hidden states
    outputs = model(**inputs)

    # outputs[0] contains the hidden states (context vectors) for each token
    hidden_states = outputs[0]

    # Tokenize the phrase separately to know how many tokens it contains
    phrase_tokens = tokenizer.encode("Quantum information theory")

    # Find the starting index of "Quantum information theory" in the tokenized sentence
    start_index = inputs['input_ids'][0].tolist().index(phrase_tokens[0])

    # Extract the context vectors for the phrase "Quantum information theory"
    phrase_context_vectors = hidden_states[0, start_index : start_index + len(phrase_tokens), :]

    # Compute pairwise Euclidean distances and store in a matrix
    distance_matrix = torch.zeros(len(phrase_tokens), len(phrase_tokens))
    for i in range(len(phrase_tokens)):
        for j in range(len(phrase_tokens)):
            distance_matrix[i, j] = euclidean_distance(phrase_context_vectors[i], phrase_context_vectors[j])

    print(f'Distance matrix for "Quantum information theory" in sentence "{sentence}":\n {distance_matrix}\n')

    # Perform PCA to reduce the dimensions to 3
    pca = PCA(n_components=3)
    reduced_vectors = pca.fit_transform(phrase_context_vectors.detach().numpy())
    
    # Create labels for each token
    labels = [tokenizer.decode([token_id]) for token_id in phrase_tokens]
    
    # Create a 3D scatter plot
    fig = go.Figure(data=[go.Scatter3d(
        x=reduced_vectors[:, 0],
        y=reduced_vectors[:, 1],
        z=reduced_vectors[:, 2],
        mode='markers+text',
        text=labels,
        marker=dict(
            size=8,
            color=distance_matrix.mean(dim=1).detach().numpy(), # set color to average distance
            colorscale='Viridis',   # choose a colorscale
            opacity=0.8
        ),
        textposition="bottom center"
    )])

    fig.update_layout(scene=dict(
                    xaxis_title='PC1',
                    yaxis_title='PC2',
                    zaxis_title='PC3'),
                    width=700,
                    margin=dict(r=20, b=10, l=10, t=10))
    fig.show()


Distance matrix for "Quantum information theory" in sentence "Quantum information theory is interesting.":
 tensor([[  0.0000, 119.8416, 172.8208, 166.4057],
        [119.8416,   0.0000,  58.0346,  55.0810],
        [172.8208,  58.0346,   0.0000,  25.0608],
        [166.4057,  55.0810,  25.0608,   0.0000]], grad_fn=<CopySlices>)



Distance matrix for "Quantum information theory" in sentence "Quantum information theory allows us to study attention using entanglement.":
 tensor([[  0.0000, 119.8416, 172.8207, 166.4057],
        [119.8416,   0.0000,  58.0346,  55.0810],
        [172.8207,  58.0346,   0.0000,  25.0608],
        [166.4057,  55.0810,  25.0608,   0.0000]], grad_fn=<CopySlices>)



In [17]:
import torch
from transformers import GPT2Tokenizer, GPT2Model
from sklearn.decomposition import PCA
import plotly.graph_objects as go

# Initialize the model and tokenizer
tokenizer = GPT2Tokenizer.from_pretrained('gpt2')
model = GPT2Model.from_pretrained('gpt2')

# Function to calculate Euclidean distance
def euclidean_distance(a, b):
    return torch.dist(a, b)

# Prompt user to enter two sentences
sentences = [input("Enter first sentence: "), input("Enter second sentence: ")]

# Iterate over sentences
for sentence in sentences:
    # Tokenize the sentence
    inputs = tokenizer(sentence, return_tensors='pt')

    # Pass the tokens through the model to get hidden states
    outputs = model(**inputs)

    # outputs[0] contains the hidden states (context vectors) for each token
    hidden_states = outputs[0]

    # Tokenize the phrase separately to know how many tokens it contains
    phrase_tokens = tokenizer.encode("Quantum information theory")

    # Find the starting index of "Quantum information theory" in the tokenized sentence
    start_index = inputs['input_ids'][0].tolist().index(phrase_tokens[0])

    # Extract the context vectors for the phrase "Quantum information theory"
    phrase_context_vectors = hidden_states[0, start_index : start_index + len(phrase_tokens), :]

    # Compute pairwise Euclidean distances and store in a matrix
    distance_matrix = torch.zeros(len(phrase_tokens), len(phrase_tokens))
    for i in range(len(phrase_tokens)):
        for j in range(len(phrase_tokens)):
            distance_matrix[i, j] = euclidean_distance(phrase_context_vectors[i], phrase_context_vectors[j])

    print(f'Distance matrix for "Quantum information theory" in sentence "{sentence}":\n {distance_matrix}\n')

    # Perform PCA to reduce the dimensions to 3
    pca = PCA(n_components=3)
    reduced_vectors = pca.fit_transform(phrase_context_vectors.detach().numpy())
    
    # Create labels for each token
    labels = [tokenizer.decode([token_id]) for token_id in phrase_tokens]
    
    # Create a 3D scatter plot
    fig = go.Figure(data=[go.Scatter3d(
        x=reduced_vectors[:, 0],
        y=reduced_vectors[:, 1],
        z=reduced_vectors[:, 2],
        mode='markers+text',
        text=labels,
        marker=dict(
            size=8,
            color=distance_matrix.mean(dim=1).detach().numpy(), # set color to average distance
            colorscale='Viridis',   # choose a colorscale
            opacity=0.8
        ),
        textposition="bottom center"
    )])

    fig.update_layout(scene=dict(
                    xaxis_title='PC1',
                    yaxis_title='PC2',
                    zaxis_title='PC3'),
                    width=700,
                    margin=dict(r=20, b=10, l=10, t=10))
    fig.show()


Distance matrix for "Quantum information theory" in sentence "Quantum information theory is an intricate and complicated topic.":
 tensor([[  0.0000, 119.8416, 172.8207, 166.4057],
        [119.8416,   0.0000,  58.0346,  55.0810],
        [172.8207,  58.0346,   0.0000,  25.0608],
        [166.4057,  55.0810,  25.0608,   0.0000]], grad_fn=<CopySlices>)



Distance matrix for "Quantum information theory" in sentence "Quantum information theory is closely related to deep learning.":
 tensor([[  0.0000, 119.8416, 172.8207, 166.4057],
        [119.8416,   0.0000,  58.0346,  55.0810],
        [172.8207,  58.0346,   0.0000,  25.0608],
        [166.4057,  55.0810,  25.0608,   0.0000]], grad_fn=<CopySlices>)



Interestingly, the distance matrices remain equal in all four conntexts computed above. That is for the sentences, 

- "Quantum information theory is interesting."
- "Quantum information theory allows us to study attention using entanglement."
- "Quantum information theory is an intricate and complicated topic."
- "Quantum information theory is closely related to deep learning."

all return the distance matrix

$$
\begin{pmatrix}
0.0000 & 119.8416 & 172.8207 & 166.4057\\
119.8416 & 0.0000 & 58.0346 & 55.0810\\
172.8207 & 58.0346 & 0.0000 & 25.0608\\
166.4057 & 55.0810 & 25.0608 & 0.0000
\end{pmatrix}
$$

Implying the relationship between the context vectors for the tokens correspodning to the phrase "Quantum information theory", that is `quant`, `um`, `information`, and `theory`, remains the same in all four contexts. What are the linguistic implications, and the implications for expressivity of GPT-2? 

When certain tokens consistently appear together in different contexts, it suggests a strong association or semantic relationship between those tokens. By maintaining a consistent representation of the pairwise distances between the context vectors of these tokens, a language model can capture and encode this association. This consistent representation allows the model to recognize the shared characteristics or semantic connections between the tokens, regardless of the specific context in which they appear.

This consistent representation of pairwise distances enables the language model to generalize the relationships between tokens and learn robust representations that capture their shared meaning or contextual dependencies. It allows the model to leverage the knowledge gained from one context to inform its understanding and generation of those tokens in different contexts.

By having a consistent representation, the model can establish a reliable and stable understanding of the associations between tokens, which can benefit downstream tasks such as information retrieval, question-answering, and text generation. It enables the model to make more accurate predictions and generate coherent and contextually appropriate responses.

However, it's important to note that while a consistent representation of pairwise distances can be beneficial for capturing common associations between tokens, it may also limit the model's ability to capture subtle variations or context-specific relationships. Depending on the task and the desired level of granularity, the trade-off between consistency and flexibility needs to be carefully considered when designing and evaluating language models.

Now, let's look at Bert. 

In [19]:
import torch
from transformers import BertTokenizer, BertModel
from sklearn.decomposition import PCA
import plotly.graph_objects as go

# Initialize the model and tokenizer
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
model = BertModel.from_pretrained('bert-base-uncased')

# Function to calculate Euclidean distance
def euclidean_distance(a, b):
    return torch.dist(a, b)

# Prompt user to enter two sentences
sentences = [input("Enter first sentence: "), input("Enter second sentence: ")]

# Iterate over sentences
for sentence in sentences:
    # Tokenize the sentence
    inputs = tokenizer(sentence, return_tensors='pt')

    # Pass the tokens through the model to get hidden states
    outputs = model(**inputs)

    # outputs.last_hidden_state contains the hidden states (context vectors) for each token
    hidden_states = outputs.last_hidden_state

    # Tokenize the phrase separately to know how many tokens it contains
    phrase_tokens = tokenizer.encode("Quantum information theory", add_special_tokens=False)

    # Find the starting index of "Quantum information theory" in the tokenized sentence
    input_ids = inputs['input_ids'].tolist()[0]
    start_index = [i for i, val in enumerate(input_ids) if val == phrase_tokens[0]][0]

    # Extract the context vectors for the phrase "Quantum information theory"
    phrase_context_vectors = hidden_states[0, start_index : start_index + len(phrase_tokens), :]

    # Compute pairwise Euclidean distances and store in a matrix
    distance_matrix = torch.zeros(len(phrase_tokens), len(phrase_tokens))
    for i in range(len(phrase_tokens)):
        for j in range(len(phrase_tokens)):
            distance_matrix[i, j] = euclidean_distance(phrase_context_vectors[i], phrase_context_vectors[j])

    print(f'Distance matrix for "Quantum information theory" in sentence "{sentence}":\n {distance_matrix}\n')

    # Perform PCA to reduce the dimensions to 3
    pca = PCA(n_components=3)
    reduced_vectors = pca.fit_transform(phrase_context_vectors.detach().numpy())

    # Create labels for each token
    labels = [tokenizer.decode(token_id) for token_id in phrase_tokens]

    # Create a 3D scatter plot
    fig = go.Figure(data=[go.Scatter3d(
        x=reduced_vectors[:, 0],
        y=reduced_vectors[:, 1],
        z=reduced_vectors[:, 2],
        mode='markers+text',
        text=labels,
        marker=dict(
            size=8,
            color=distance_matrix.mean(dim=1).detach().numpy(),  # set color to average distance
            colorscale='Viridis',  # choose a colorscale
            opacity=0.8
        ),
        textposition="bottom center"
    )])

    fig.update_layout(scene=dict(
        xaxis_title='PC1',
        yaxis_title='PC2',
        zaxis_title='PC3'),
        width=700,
        margin=dict(r=20, b=10, l=10, t=10))
    fig.show()


Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertModel: ['cls.predictions.transform.dense.bias', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.transform.dense.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.bias']
- This IS expected if you are initializing BertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


Distance matrix for "Quantum information theory" in sentence "Quantum information theory is an intricate and complicated topic.":
 tensor([[ 0.0000, 14.8998, 20.3941],
        [14.8998,  0.0000, 16.3631],
        [20.3941, 16.3631,  0.0000]], grad_fn=<CopySlices>)



Distance matrix for "Quantum information theory" in sentence "Quantum information theory is closely related to deep learning.":
 tensor([[ 0.0000, 18.2120, 21.2747],
        [18.2120,  0.0000, 16.7886],
        [21.2747, 16.7886,  0.0000]], grad_fn=<CopySlices>)



As we can see, the Bert model does not maintain the pairwise distances between the tokens' context vectors, and we get two different distance matrices. The varying pairwise distances observed in BERT's context vectors for the phrase "Quantum information theory" across different sentences indicate that BERT does not encode the tokens as a unique conceptual unit.

BERT's contextualized representations are designed to capture fine-grained contextual information and encode the relationships between tokens based on their specific context. In this case, the varying distances suggest that BERT differentiates the context vectors for each token in the phrase based on the surrounding words and their specific contextual associations.